biomage/stage/stage.py

```diff
@@ -224,12 +257,284 @@ def create_manifest(templates, token):
+    click.echo()
+    click.echo(
+        click.style("Create isolated staging environment.", fg="yellow", bold=True)
+    )
+    click.echo(
+        "To provide isolation, files and records from existing experimentIds "
+        "will be copied and renamed under unique experimentIds.\n"
+        "You can use these scoped resources to test your changes. "
+        "Be mindful, that creating isolated environments create resources in AWS."
+    )
+
+    staged_experiments = [
+        experiment
+        for experiment in all_experiments
+        if re.match(f"^{sandbox_id}-*", experiment["experimentId"])
+    ]
+
+    # Exclude experiments with the pattern {sandbox_id}-{experiment_id}
+    # and the associated experiment_id to prevent double staging.
+    excluded_experiment_ids = [
+        experiment["experimentId"].replace(f"{sandbox_id}-", "")
+        for experiment in staged_experiments
+    ] + [experiment["experimentId"] for experiment in staged_experiments]
+
+    unstaged_experiments = [
+        experiment
+        for experiment in all_experiments
+        if experiment["experimentId"] not in excluded_experiment_ids
+    ]
+
+    if len(staged_experiments) > 0:
+        click.echo()
+        click.echo(f"Staged environment(s) exists for the sandbox_id {sandbox_id}:")
+        click.echo(
+            "\n".join(
+                [f"• {experiment['experimentId']}" for experiment in staged_experiments]
+            )
+        )
+        click.echo()
+
+    # Implement pagination if result contains more than experiments_per_page number of experiments
+
+    chosen_experiments = []
+    experiments_per_page = 10
+
+    if len(unstaged_experiments) > experiments_per_page:
+
+        max_page = math.ceil(len(unstaged_experiments) / experiments_per_page)
+        page = 0
+        done = False
+        previous_text = "<< previous <<"
+        next_text = ">> next >>"
+        done_text = "== done =="
+
+        while not done:
+
+            idx_start = experiments_per_page * page
+            idx_end = (page + 1) * experiments_per_page
+
+            choices = [
+                {
+                    "name": "{} {}".format(
+                        experiment["experimentId"].ljust(40),
+                        experiment["experimentName"],
+                    ),
+                    "checked": experiment["experimentId"] in chosen_experiments,
+                }
+                for experiment in unstaged_experiments[idx_start:idx_end]
+            ]
+
+            if page > 0:
+                choices = [{"name": previous_text}] + choices
+
+            if page < max_page - 1:
+                choices += [{"name": next_text}]
+
+            choices += [{"name": done_text}]
+
+            questions = [
+                {
+                    "type": "checkbox",
+                    "name": "experiments_to_stage",
+                    "message": "Which experiments would you like to enable for the staging environment?\n"
```

Comment:
I see what you are doing here and there is certainly going to be a need for some pagination when we have more and more experiments in staging. However, the code is quite difficult to read at this point -- there is too much pagination-specific logic in that `choose_staging_experiments` function, which is about 200 lines long. What could help I am thinking is to extract the logic from lines 313 to 407 into a separate private function that will take care of pagination and then just use that function in here. What do you think? Do you have any other ideas?

Reply:
Also, just to keep in mind that in the not so long future, we will be creating staging environments programmatically from CI

Reply:
I can extract the pagination out into a function.

Regarding creation of scoped experiments in CI, I think we can pass `experiment_ids` into `wf.create_dispatch()` https://github.com/biomage-ltd/biomage-utils/blob/master/biomage/stage/stage.py#L299-L302 and create custom actions to the setting up of the environment as a Github action.

Reply:
yeah looks like it will work indeed